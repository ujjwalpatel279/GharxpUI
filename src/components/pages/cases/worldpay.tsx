import React, { FunctionComponent, ReactElement, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { Container } from '../../shared';
import { services } from '../../../shared';

interface UrlParamInterface {
  brokerGuid: string;
  uniqueOrderCode?: string;
  applicationFeeSummaryId?: string;
}

export const WorldPayPage: FunctionComponent = (): ReactElement => {
  const { uniqueOrderCode, applicationFeeSummaryId } = useParams<UrlParamInterface>();
  const fetchWorldPayResponse = async () => {
    await services.worldPayOrderResponse(uniqueOrderCode, applicationFeeSummaryId, location.search);
  };
  useEffect(() => {
    if ((uniqueOrderCode ?? '') === '' && (applicationFeeSummaryId ?? '') === '') {
      location.href = location.search.substr(1, location.search.length - 1);
    } else {
      fetchWorldPayResponse();
    }
  }, []);

  return (
    <section>
      <Container>
        {(uniqueOrderCode ?? '') === '' && (applicationFeeSummaryId ?? '') === '' && (
          <div>Loading Worldpay Redirection Url</div>
        )}
        {(uniqueOrderCode ?? '') !== '' && (applicationFeeSummaryId ?? '') !== '' && (
          <div>Processing completed. Please shut down this window</div>
        )}
      </Container>
    </section>
  );
};
