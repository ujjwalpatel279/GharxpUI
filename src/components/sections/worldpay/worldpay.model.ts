export interface CustomerDocumentsInterface {
  canBeDeleted?: boolean;
  compressed?: boolean;
  customerId: number;
  description: string;
  docCode?: string;
  document?: Int32Array;
  formId?: number;
  id: number;
  mimeType?: string;
  typeCatagory?: string;
  typeId?: number;
  uploadedOn?: string;
}
