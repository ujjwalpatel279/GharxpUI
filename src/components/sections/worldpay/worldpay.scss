.c-add-fee {
  .modal {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.8);
    z-index: 20;
  }

  input[type='file'] {
    margin-bottom: 0.375rem !important;
  }

  .input-type-hidden {
    display: none;
  }

  .continue-button {
    margin-top: 10px;
    float: right;
  }

  .modal-main {
    position: fixed;
    background: white;
    width: 70%;
    height: 70%;
    top: 30%;
    left: 30%;
    transform: translate(-50%, -50%);
    padding: 20px;
    overflow: auto;
  }

  .upload-button {
    float: right;
  }

  .c-form-input {
    width: 100%;
  }

  .display-block {
    display: block;
  }

  .display-none {
    display: none;
  }
}
