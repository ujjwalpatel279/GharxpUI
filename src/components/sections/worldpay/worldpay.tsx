import React, { FunctionComponent, useEffect, useState } from 'react';
import { Button, FormField, Heading, Input, InputList } from '../../shared';
import { useForm } from 'react-hook-form';

import { defaultErrorResponse, hasError, ServiceHandlingInterface } from '../../../shared/types/service-handling';

import './worldpay.scss';

import {
  FeeInterface,
  ListItemInterface,
  WorldPayOrderInterface,
  WorldPayResponseInterface,
} from '../../../shared/models';
import { services } from '../../../shared';

export interface WorldPayInterface {
  handleClose: () => void;
  show: boolean;
  fees: FeeInterface[];
  reloadGrids?: (caseId: number, formId: number) => void;
  caseId: number;
  formId: number;
}

type Inputs = {
  emailAddress: string;
  addressLine1: string;
  city: string;
  postcode: string;
  countryCode: string;
  reference: string;
};

export const worldPaySubmitInitialState: ServiceHandlingInterface<WorldPayResponseInterface> = {
  data: undefined,
  loading: false,
  error: false,
};

export const WorldPay: FunctionComponent<WorldPayInterface> = ({
  handleClose,
  show,
  fees,
  reloadGrids,
  caseId,
  formId,
}) => {
  const [worldPayError, setWorldPayError] = useState<string>('');
  const [worldPayWindow, setWorldPayWindow] = useState<Window>(null);
  const [connection, setConnection] = useState(null);
  const [shWorldPaySubmit, setShWorldPaySubmit] = useState<ServiceHandlingInterface<WorldPayResponseInterface>>(
    worldPaySubmitInitialState,
  );

  const { register, handleSubmit, setValue, errors } = useForm<Inputs>();
  const showHideClassName = show ? 'modal display-block' : 'modal display-none';
  const countries: ListItemInterface[] = [
    {
      id: -1,
      value: 'GB',
      code: 'GB',
      displayText: 'United Kingdom',
    },
    {
      id: -2,
      value: 'US',
      code: 'US',
      displayText: 'United States',
    },
  ];

  const validation = {
    emailAddress: {
      types: { required: true, pattern: /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,4}$/ },
      message: `Please enter an email address`,
    },
    addressLine1: {
      types: { required: true },
      message: `Please enter a billing address`,
    },
    city: {
      types: { required: true },
      message: `Please enter a city`,
    },
    postcode: {
      types: { required: true },
      message: `Please enter a postcode`,
    },
    countryCode: {
      types: { required: true },
      message: `Please enter a country`,
    },
    reference: {
      types: { required: true },
      message: `Please enter a reference`,
    },
  };

  const startCheckingForPaymentStatus = async (worldPayOrder: WorldPayOrderInterface): Promise<string> => {
    setShWorldPaySubmit({ ...shWorldPaySubmit, loading: true });
    const response = await services.submitWorldPayOrder(caseId, formId, worldPayOrder);
    setShWorldPaySubmit(response);

    if (response.data && response.data.isOrderSucceeded && (response.data.redirectionUrl ?? '').length > 0) {
      const newConnection = services.getHubConnection();
      setConnection(newConnection);
    }
    return response.data.isOrderSucceeded ? response.data.redirectionUrl : '';
  };

  const sendToWorldPay = async (data: Inputs) => {
    const worldPayOrder: WorldPayOrderInterface = {
      emailAddress: data.emailAddress,
      addressLine1: data.addressLine1,
      city: data.city,
      postcode: data.postcode,
      countryCode: data.countryCode,
      reference: data.reference,
      fees: fees,
      amount: fees.map((fee) => fee.amount).reduce((acc, next) => (acc ?? 0) + next),
    };

    const redirectUrl = await startCheckingForPaymentStatus(worldPayOrder);
    if ((redirectUrl ?? '').length > 0) {
      const wpw = window.open(`/worldpay/?${redirectUrl}`, '_blank');
      setWorldPayWindow(wpw);
      setWorldPayError('Processing worldpay payment...');
    } else {
      setWorldPayError('Payment could not be processed');
    }
  };

  useEffect(() => {
    if (connection) {
      connection.start().then(() => {
        connection.invoke('getConnectionId').then(function (connectionId: string) {
          // Send the connectionId to controller
        });
        connection.on('paymentProcessed', (message: string) => {
          if (worldPayWindow !== null) {
            worldPayWindow.close();
          }
          reloadGrids(caseId, formId);
          handleClose();
          connection.stop();
        });
      });
    }
  }, [connection]);

  if (hasError([shWorldPaySubmit])) {
    return defaultErrorResponse();
  }

  return (
    <div className={showHideClassName}>
      <section className="modal-main">
        <Heading level="h3" title="Client Contact Details" />
        <br />

        <form onSubmit={handleSubmit(sendToWorldPay)}>
          {worldPayError && (
            <p role="alert" className="c-form-field__error">
              {worldPayError}
            </p>
          )}
          <FormField label="Email Address" error={errors.emailAddress && validation.emailAddress.message}>
            <Input
              id="emailAddress"
              placeHolder="Email address"
              error={!!errors.emailAddress}
              ref={register(validation.emailAddress.types)}
            />
          </FormField>
          <FormField label="Address Line 1" error={errors.addressLine1 && validation.addressLine1.message}>
            <Input
              id="addressLine1"
              placeHolder="Address Line 1"
              error={!!errors.addressLine1}
              ref={register(validation.addressLine1.types)}
            />
          </FormField>
          <FormField label="City" error={errors.city && validation.city.message}>
            <Input id="city" placeHolder="City" error={!!errors.city} ref={register(validation.city.types)} />
          </FormField>
          <FormField label="Postcode" error={errors.postcode && validation.postcode.message}>
            <Input
              id="postcode"
              placeHolder="Postcode"
              error={!!errors.postcode}
              ref={register(validation.postcode.types)}
            />
          </FormField>
          <FormField label="Country" error={errors.countryCode && validation.countryCode.message}>
            <InputList
              id="countryCode"
              options={countries}
              legend="Select a country"
              forceSelection
              error={!!errors.countryCode}
              ref={register(validation.countryCode.types)}
            />
          </FormField>
          <FormField label="Reference" error={errors.reference && validation.reference.message}>
            <Input
              id="reference"
              placeHolder="Reference"
              error={!!errors.reference}
              ref={register(validation.reference.types)}
            />
          </FormField>
          <Button type="button" variant="navy" size="small" onClick={handleClose}>
            Cancel
          </Button>
          <Button type="submit" childClasses="upload-button" variant="primary" size="small">
            Pay
          </Button>
        </form>
      </section>
    </div>
  );
};
