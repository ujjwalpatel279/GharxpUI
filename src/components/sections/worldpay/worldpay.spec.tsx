import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { WorldPay } from './worldpay';

describe('WorldPay Component', () => {
  afterEach(cleanup);

  it('renders the WorldPay component', () => {
    const { container } = render(<WorldPay show={true} fees={[]} handleClose={() => null} caseId={1} formId={1} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Shows correct form text', () => {
    const { getByText } = render(
      <BrowserRouter>
        <WorldPay show={true} fees={[]} handleClose={() => null} caseId={1} formId={1} />
      </BrowserRouter>,
    );

    expect(getByText('Email Address')).toBeInTheDocument();
    expect(getByText('Address Line 1')).toBeInTheDocument();
    expect(getByText('Postcode')).toBeInTheDocument();
    expect(getByText('Country')).toBeInTheDocument();
    expect(getByText('Cancel')).toBeInTheDocument();
    expect(getByText('Pay')).toBeInTheDocument();
  });
});
