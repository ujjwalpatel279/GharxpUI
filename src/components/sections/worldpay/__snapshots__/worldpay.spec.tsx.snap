// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorldPay Component renders the WorldPay component 1`] = `
<div
  class="modal display-block"
>
  <section
    class="modal-main"
  >
    <h3
      class="c-heading c-heading--h3"
    >
      Client Contact Details
    </h3>
    <br />
    <form>
      
      <div
        class="c-form-field"
      >
        <label
          class="c-form-field__label"
        >
          Email Address
        </label>
        <input
          aria-invalid="false"
          class="c-input"
          id="emailAddress"
          name="emailAddress"
          placeholder="Email address"
          type="text"
        />
      </div>
      <div
        class="c-form-field"
      >
        <label
          class="c-form-field__label"
        >
          Address Line 1
        </label>
        <input
          aria-invalid="false"
          class="c-input"
          id="addressLine1"
          name="addressLine1"
          placeholder="Address Line 1"
          type="text"
        />
      </div>
      <div
        class="c-form-field"
      >
        <label
          class="c-form-field__label"
        >
          City
        </label>
        <input
          aria-invalid="false"
          class="c-input"
          id="city"
          name="city"
          placeholder="City"
          type="text"
        />
      </div>
      <div
        class="c-form-field"
      >
        <label
          class="c-form-field__label"
        >
          Postcode
        </label>
        <input
          aria-invalid="false"
          class="c-input"
          id="postcode"
          name="postcode"
          placeholder="Postcode"
          type="text"
        />
      </div>
      <div
        class="c-form-field"
      >
        <label
          class="c-form-field__label"
        >
          Country
        </label>
        <fieldset>
          <legend>
            Select a country
          </legend>
          <div
            class="l-flex l-flex--wrap"
          >
            <div>
              <label
                class="c-input-list-item"
                for="-1"
              >
                <input
                  id="-1"
                  name="countryCode"
                  type="radio"
                  value="GB"
                />
                <span
                  class="c-radio-input__inner"
                >
                  <span>
                    GB
                  </span>
                </span>
              </label>
            </div>
            <div>
              <label
                class="c-input-list-item"
                for="-2"
              >
                <input
                  id="-2"
                  name="countryCode"
                  type="radio"
                  value="US"
                />
                <span
                  class="c-radio-input__inner"
                >
                  <span>
                    US
                  </span>
                </span>
              </label>
            </div>
          </div>
        </fieldset>
      </div>
      <div
        class="c-form-field"
      >
        <label
          class="c-form-field__label"
        >
          Reference
        </label>
        <input
          aria-invalid="false"
          class="c-input"
          id="reference"
          name="reference"
          placeholder="Reference"
          type="text"
        />
      </div>
      <button
        class="c-btn c-btn--navy c-btn--small"
        type="button"
      >
        <span>
          Cancel
        </span>
      </button>
      <button
        class="c-btn c-btn--primary c-btn--small upload-button"
        type="submit"
      >
        <span>
          Pay
        </span>
      </button>
    </form>
  </section>
</div>
`;
